### PR TITLE
Remove gap, proposals, vote, tally

### PR DESCRIPTION
**Pull request by thanhnguyennguyen**
- Calculating set of masternodes at checkpoint, remove gap, proposal, votes, tally.
- Testing scenarios:
   - Updating the set of masternodes (turn off some masternodes and turn it on after a few epochs)
  - Full nodes / masternodes

### CODE REVIEW

**Comment by thanhnguyennguyen** [comment]
fix #213

**Comment by thanhnguyennguyen** [comment]
Hi @ngtuna 
Could you please arrange your time help me review this PR ?
Thanks,
NguyenNguyen

File: consensus/posv/posv.go [hunk 1]
```diff
@@ -792,8 +770,34 @@ func (c *Posv) Prepare(chain consensus.ChainReader, header *types.Header) error
...
+			if err != nil {
+				return err
+			}
+			for _, m := range ms {
+				snap.Signers[m.Address] = struct{}{}
+			}
+			c.recents.Add(snap.Hash, snap)
+			log.Info("New set of masternodes has been updated to snapshot", "number", snap.Number, "hash", snap.Hash)
+		}
+		masternodes := snap.GetSigners()
```
Comment (nguyenbatam) [inline consensus/posv/posv.go]: Code luc truoc : verify & insert xong block 900 thi moi updateM1 voi block hash 900
Code cho nay thi moi dang verify block 900 & chua insert xong thi update M1 , voi block 899

File: consensus/posv/posv.go [hunk 2]
```diff
@@ -391,19 +373,40 @@ func (c *Posv) verifyCascadingFields(chain consensus.ChainReader, header *types.
...
+			ms, err := c.HookUpdateMNs()
+			if err != nil {
+				return err
+			}
+			for _, m := range ms {
+				snap.Signers[m.Address] = struct{}{}
+			}
+			// reset
+			c.checkpointHooked = map[uint64]bool{}
+			c.checkpointHooked[number] = true
```
Comment (nguyenbatam) [inline consensus/posv/posv.go]: map checkpointHooked la thua , Chain se ko verify hay insert 1 block cung hash 
Ma truong hop nay la gap 2 block 900 nhung khac hash vi the snapshot tuong ung voi moi block hash nay deu can phai tinh lai

File: params/config.go [hunk 3]
```diff
@@ -149,7 +149,6 @@ type PosvConfig struct {
 	Epoch               uint64         `json:"epoch"`               // Epoch length to reset votes and checkpoint
 	Reward              uint64         `json:"reward"`              // Block reward - unit Ether
 	RewardCheckpoint    uint64         `json:"rewardCheckpoint"`    // Checkpoint block for calculate rewards.
-	Gap                 uint64         `json:"gap"`                 // Gap time preparing for the next epoch
```
Comment (nguyenbatam) [inline params/config.go]: Cai Gap  nay thi giu lai di de danh cho hardfork

File: consensus/posv/posv.go [hunk 4]
```diff
@@ -533,22 +523,52 @@ func (c *Posv) YourTurn(chain consensus.ChainReader, parent *types.Header, signe
...
+	}
+	snap, err := c.snapshot(chain, number, header.Hash(), nil)
+	if err != nil {
+		return err
+	}
+	newMasternodes := make(map[common.Address]struct{})
+	for i:= 0; i < len(masternodes); i++ {
+		newMasternodes[masternodes[i]] = struct{}{}
+	}
+	snap.Signers = newMasternodes
```
Comment (nguyenbatam) [inline consensus/posv/posv.go]: Truoc : map Signers nay chua top 150 candidate
Sau : map Signers nay chi chua list master nodes =>  trong phan tao chu ki co 1 dieu kien  check tao chu ki voi map nay =>  nhung node bi slash se ko thuc hien ki nen 1 trong 2 se phai sua them 
Function HookUpdateMNs goi khi Prepare & verifyHeader => thoi diem chot so la 899 , 
O  day duoc goi sau khi insert nen da chuyen sang block 900 neu goi lai thi co the bi lech

Comment (nguyenbatam) [reply consensus/posv/posv.go]: Co le nen do lai 4,5 epoch de lay list bi slashed add them vao

File: consensus/posv/posv.go [hunk 5]
```diff
@@ -533,22 +517,50 @@ func (c *Posv) YourTurn(chain consensus.ChainReader, parent *types.Header, signe
...
+	if err != nil {
+		return []common.Address{}, err
+	}
+	var masternodes []common.Address
+	// reset signer list
+	snap.Signers = map[common.Address]struct{}{}
+	for _, m := range ms {
+		masternodes = append(masternodes, m.Address)
+		snap.Signers[m.Address] = struct{}{}
+	}
```
Comment (nguyenbatam) [inline consensus/posv/posv.go:551]: Viet lai ntn de tranh bug concurrent map read and map write
    newMasternodes := make(map[common.Address]struct{})
	for _, m := range ms {
		newMasternodes[m.Address] = struct{}{}
	}
	snap.Signers = newMasternodes

Comment (thanhnguyennguyen) [reply consensus/posv/posv.go:551]: :+1:

**Comment by thanhson1085** [comment]
Will open back if have issue

**Pull request closed without merging**